### ISSUE TITLE
Broken Quantity type after upgrade to 2.3

### ISSUE THREAD

**Issue by alex-mashin** [SemanticMediaWiki/SemanticMediaWiki#1529]
After upgrade to SMW 2.3 under MW 1.26.2, MySQL storage (MariaDB) my currency properties stopped working.

I have several of them, i.e. https://traditio.wiki/%D0%A1%D0%B2%D0%BE%D0%B9%D1%81%D1%82%D0%B2%D0%BE:%D0%92%D1%8B%D1%80%D1%83%D1%87%D0%BA%D0%B0 (in Russian, "Property:Revenue"). Their definitions include the template https://traditio.wiki/%D0%A8%D0%B0%D0%B1%D0%BB%D0%BE%D0%BD:%D0%98%D0%B7%D0%BC%D0%B5%D1%80%D1%8F%D0%B5%D1%82%D1%81%D1%8F_%D0%B4%D0%B5%D0%BD%D1%8C%D0%B3%D0%B0%D0%BC%D0%B8 ("Template:Measured as money"), which defines units of measurement for several currencies (with exchange rates as of January 2016) and also thousands, millions and billions of them.

It worked fine in January 2016, under SMW 2.2 until I decided to upgrade.

Now it doesn't. Pages with semantic annotations with currency properties and units of measurement that I have defined for them now appear to have improper values. Example: https://traditio.wiki/%D0%A2%D1%80%D0%B0%D0%B4%D0%B8%D1%86%D0%B8%D1%8F:%D0%A2%D0%B5%D1%81%D1%822. Units that have 1:1 ratio (руб., ₽) are recognised; others aren't ($, USD, тыс. руб.).

It seems that you have broken the special property "Corresponds to".

I have also noticed that spaces are now missing from units of measurement set with "Display units" property (the same example; "млн" and "руб." should have ". " between them but don't.

**Comment by mwjames** [comment]
> It worked fine in January 2016, under SMW 2.2 until I decided to upgrade

There have been no changes to either Quantity or Unit (\[0]) processing (as far as I remember) between 2.1 and 2.3.1 [0] therefore I cannot validate the claim which rests on the 2.2 vs 2.3 argument. We do have changes that were made in the upcoming 2.4 release but those were not deployed with 2.3*.

> It seems that you have broken the special property "Corresponds to".

We have different regression tests [1, 2, 3, 4, 5] that include different conversion rules in connection with currency and "Corresponds to" and they did not indicate malfunction or deviation from the expected output.

> I have also noticed that spaces are now missing from units of measurement set with "Display units" property (the same example; "млн" and "руб." should have ". " between them but don't.

I'm not sure I understand the issue and as noted before please provide an example on the [sandbox](http://sandbox.semantic-mediawiki.org).

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/commits/master/includes/datavalues/SMW_DV_Quantity.php

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/ByJsonScript/Fixtures/q-0502.json
[2] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/ByJsonScript/Fixtures/q-0501.json
[3] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/ByJsonScript/Fixtures/q-0503.json
[4] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/ByJsonScript/Fixtures/f-0204.json
[5] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/tests/phpunit/Integration/ByJsonScript/Fixtures/p-0209.json

PS: As previously noted (in the ISSUE_TEMPLATE), please provide an example on the sandbox to isolate an example and described your observation so it is easily identifiable and if necessary be included as regression tests against future changes.

PSS: If no sufficient examples are provided then it cannot be expected that a developer(in) will provide any assistance within its volunteer capacity.

**Comment by alex-mashin** [comment]
As you wish.

Template: http://sandbox.semantic-mediawiki.org/wiki/Mod%C3%A8le:Monetary.
Property: http://sandbox.semantic-mediawiki.org/wiki/Attribut:Has_revenue.
Page with property: http://sandbox.semantic-mediawiki.org/wiki/Zorg_holdings.

**Comment by mwjames** [comment]
> As you wish.

Yes, the examples helped to find an explanation on what is happing (and luckily the sandbox works with a different content language FR).

The observation relates to separators specified by the content language are used by the `NumberValue` regex to interpret decimal numeric values.

```
$decseparator = NumberFormatter::getInstance()->getDecimalSeparatorForContentLanguage();
$kiloseparator = NumberFormatter::getInstance()->getThousandsSeparatorForContentLanguage();
```
```
		$regex = '/([-+]?\s*(?:' .
				// Either numbers like 10,000.99 that start with a digit
				'\d+(?:\\' . $kiloseparator . '\d\d\d)*(?:\\' . $decseparator . '\d+)?' .
				// or numbers like .001 that start with the decimal separator
				'|\\' . $decseparator . '\d+' .
				')\s*(?:[eE][-+]?\d+)?)/u';
```
If you where switch to EN as content language the interpretation of `.` would allow to parse `0.013711778417661 $` as decimal punctuation while in case of the sandbox (== FR) `.` is not recognized as a valid decimal separator (`,` is expected instead).

**Comment by mwjames** [comment]
> It worked fine in January 2016, under SMW 2.2 until I decided to upgrade.

The code for parsing numeric values (and hereof decimal related strings) in 2.1 [0] and 2.3 [1] is the same.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/2.1.x/includes/datavalues/SMW_DV_Number.php#L71-L78

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/2.3.x/includes/datavalues/SMW_DV_Number.php#L71-L77

Parsing output where `.` is recognized as decimal separator.
```
array(3) {
  [0]=>
  string(0) ""
  [1]=>
  string(17) "0.013711778417661"
  [2]=>
  string(39) "$,USD,dollar,dollars,USdollar,USdollars"
}
```

Parsing output where `.` was not recognized (as for FR as site/content language).
```
array(3) {
  [0]=>
  string(0) ""
  [1]=>
  string(1) "0"
  [2]=>
  string(55) ".013711778417661$,USD,dollar,dollars,USdollar,USdollars"
}
```

**Comment by alex-mashin** [comment]
Thank you, replacing `tostring ()` with `mw.getContentLanguage ():formatNum ()` helped.

I might have changed some MW language setttings while upgrading.

There is also the inner space problem: `US dollar` → `USdollar`. That `str_replace( array( '&nbsp;', '&#160;', '&thinsp;', ' ' )` replaces too much. Perhaps, something like `preg_replace ('/(?<=^|\d)(&nbsp;|&#160;|&thinsp;|\s|\p{Mc})+/u', '', $value)` will be better.

**Comment by alex-mashin** [comment]
Sorry, closed by mistake.

**Comment by mwjames** [comment]
> here is also the inner space problem: US dollar → USdollar. That str_replace( array( '&nbsp;', '&#160;', '&thinsp;', ' ' ) replaces too much.

Doing so would actually be a breaking change because things like `km²` and `km ²` would no longer return the same results for a query since internally `km²` !== `km ²`. 

The removal of spaces in current form has been at least be available since 1.8 [0] therefore changing this is not a bug but a feature request.

Tests would need adaptation to reflect this behaviour change and documentation would require to be altered in order to make sure that spaces are explicit listed as part of the specification to avoid disruption of the parsing process.

* `[[Corresponds to::1 km²]]  => [[Corresponds to::1 km², km ²]]`
* `[[Corresponds to::0.38610 sq mi]] => [[Corresponds to::0.38610  sqmi, sq mi]]`
* `[[Corresponds to::1000 m²]] => [[Corresponds to::1000 m², m ²]]`

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/1.8.0.5/includes/datavalues/SMW_DV_Number.php#L74

**Comment by mwjames** [comment]
> here is also the inner space problem: US dollar → USdollar.
> Tests would need adaptation to reflect this behaviour change and documentation would require to be altered in order to make sure that spaces are explicit listed as part of the specification to avoid disruption 

Has been addressed with #1718.
